File: test/CompatibilityHarness/java/src/main/java/com/microsoft/azure/functions/kafka/compat/CompatibilityFunctions.java
package com.microsoft.azure.functions.kafka.compat;

import com.microsoft.azure.functions.ExecutionContext;
import com.microsoft.azure.functions.HttpMethod;
import com.microsoft.azure.functions.HttpRequestMessage;
import com.microsoft.azure.functions.HttpResponseMessage;
import com.microsoft.azure.functions.HttpStatus;
import com.microsoft.azure.functions.OutputBinding;
import com.microsoft.azure.functions.annotation.AuthorizationLevel;
import com.microsoft.azure.functions.annotation.FunctionName;
import com.microsoft.azure.functions.annotation.HttpTrigger;
import com.microsoft.azure.functions.annotation.KafkaOutput;
import com.microsoft.azure.functions.annotation.KafkaTrigger;
import java.util.Optional;

public class CompatibilityFunctions {
    @FunctionName("ProduceCompatibilityMessage")
    public HttpResponseMessage produce(
            @HttpTrigger(
                    name = "request",
                    methods = {HttpMethod.POST},
                    authLevel = AuthorizationLevel.ANONYMOUS)
                    HttpRequestMessage<Optional<String>> request,
            @KafkaOutput(
                    name = "output",
                    topic = "compat-input",
                    brokerList = "kafka:29092")
                    OutputBinding<String> output,
            ExecutionContext context) {
        String message = request.getBody().orElseThrow(
                () -> new IllegalArgumentException("A correlation token is required."));
        output.setValue(message);
        context.getLogger().info("Produced compatibility token: " + message);
        return request.createResponseBuilder(HttpStatus.ACCEPTED).body(message).build();
    }

    @FunctionName("RoundTripCompatibilityMessage")
    public void roundTrip(
            @KafkaTrigger(
                    name = "input",
                    topic = "compat-input",
                    brokerList = "kafka:29092",
                    consumerGroup = "compat-java")
                    String event,
            @KafkaOutput(
                    name = "output",
                    topic = "compat-result",
                    brokerList = "kafka:29092")
                    OutputBinding<String> output,
            ExecutionContext context) {
        output.setValue(event);
        context.getLogger().info("Forwarded compatibility event.");
    }
}

File: test/CompatibilityHarness/dotnet-isolated/CompatibilityFunctions.cs
using System.Net;
using Microsoft.Azure.Functions.Worker;
using Microsoft.Azure.Functions.Worker.Http;

namespace KafkaCompatibility;

public static class CompatibilityFunctions
{
    [Function("ProduceCompatibilityMessage")]
    public static ProduceResult Produce(
        [HttpTrigger(AuthorizationLevel.Anonymous, "post")] HttpRequestData request)
    {
        using var reader = new StreamReader(request.Body);
        var message = reader.ReadToEnd();
        if (string.IsNullOrWhiteSpace(message))
        {
            throw new InvalidOperationException("A correlation token is required.");
        }

        var response = request.CreateResponse(HttpStatusCode.Accepted);
        response.WriteString(message);
        return new ProduceResult
        {
            Message = message,
            Response = response,
        };
    }

    [Function("RoundTripCompatibilityMessage")]
    [KafkaOutput("kafka:29092", "compat-result")]
    public static string RoundTrip(
        [KafkaTrigger(
            "kafka:29092",
            "compat-input",
            ConsumerGroup = "compat-dotnet-isolated")]
        string message)
    {
        return message;
    }
}

public sealed class ProduceResult
{
    [KafkaOutput("kafka:29092", "compat-input")]
    public required string Message { get; init; }

    public required HttpResponseData Response { get; init; }
}
